File: 16.Software-Design-Patterns/example/Structural/Facade/Facade.java
// This class represents a single lightbulb
 class Light {
  public void turnOn() {
    System.out.println("Light turned on");
  }

  public void turnOff() {
    System.out.println("Light turned off");
  }
}

// This class represents the complex subsystem: the LightSystem
 class LightSystem {
  private Light kitchenLight;
  private Light livingRoomLight;

  public LightSystem() {
    kitchenLight = new Light();
    livingRoomLight = new Light();
  }

  // These methods control individual lights within the system
  public void turnOnKitchenLights() {
    kitchenLight.turnOn();
  }

  public void turnOffKitchenLights() {
    kitchenLight.turnOff();
  }

  public void turnOnLivingRoomLights() {
    livingRoomLight.turnOn();
  }

  public void turnOffLivingRoomLights() {
    livingRoomLight.turnOff();
  }
}

// This class acts as the facade, simplifying interaction with the LightSystem
 class LightController {
  private LightSystem lightSystem;

  public LightController() {
    lightSystem = new LightSystem();
  }

  // These methods provide a simplified interface for controlling all lights
  public void turnOnAllLights() {
    lightSystem.turnOnKitchenLights();
    lightSystem.turnOnLivingRoomLights();
  }

  public void turnOffAllLights() {
    lightSystem.turnOffKitchenLights();
    lightSystem.turnOffLivingRoomLights();
  }
}

public class Main {
  public static void main(String[] args) {
    // Create an instance of the facade (LightController)
    LightController lightController = new LightController();

    // Use the facade's methods to control all lights at once
    lightController.turnOnAllLights(); // Output: Light turned on (twice)
    lightController.turnOffAllLights(); // Output: Light turned off (twice)
  }
}

File: 2.Intermediate Level/6.Software-Design-Patterns/example/Structural/Facade/Facade.go
package main

import (
	"fmt"
)

// Light represents a single lightbulb
type Light struct{}

// turnOn turns on the light
func (l *Light) turnOn() {
	fmt.Println("Light turned on")
}

// turnOff turns off the light
func (l *Light) turnOff() {
	fmt.Println("Light turned off")
}

// LightSystem manages multiple lights
type LightSystem struct {
	kitchenLight Light
	livingRoomLight Light
}

// NewLightSystem creates a new LightSystem
func NewLightSystem() *LightSystem {
	return &LightSystem{
		kitchenLight: Light{},
		livingRoomLight: Light{},
	}
}

// turnOnKitchenLights turns on the kitchen light
func (ls *LightSystem) turnOnKitchenLights() {
	ls.kitchenLight.turnOn()
}

// turnOffKitchenLights turns off the kitchen light
func (ls *LightSystem) turnOffKitchenLights() {
	ls.kitchenLight.turnOff()
}

// turnOnLivingRoomLights turns on the living room light
func (ls *LightSystem) turnOnLivingRoomLights() {
	ls.livingRoomLight.turnOn()
}

// turnOffLivingRoomLights turns off the living room light
func (ls *LightSystem) turnOffLivingRoomLights() {
	ls.livingRoomLight.turnOff()
}

// LightController is the facade for interacting with the LightSystem
type LightController struct {
	lightSystem *LightSystem
}

// NewLightController creates a new LightController
func NewLightController(ls *LightSystem) *LightController {
	return &LightController{
		lightSystem: ls,
	}
}

// turnOnAllLights turns on all lights
func (lc *LightController) turnOnAllLights() {
	lc.lightSystem.turnOnKitchenLights()
	lc.lightSystem.turnOnLivingRoomLights()
}

// turnOffAllLights turns off all lights
func (lc *LightController) turnOffAllLights() {
	lc.lightSystem.turnOffKitchenLights()
	lc.lightSystem.turnOffLivingRoomLights()
}

func main() {
	// Create a LightSystem
	lightSystem := NewLightSystem()

	// Create a LightController using the LightSystem
	lightController := NewLightController(lightSystem)

	// Use the LightController to manage lights
	lightController.turnOnAllLights()
	lightController.turnOffAllLights()
}
